radiometric_normalization/pif.py

```diff
@@ -72,3 +81,156 @@ def _filter_zero_alpha_pifs(reference_gimage, candidate_gimage):
     pif_weight[valid_pixels] = reference_gimage.alpha[valid_pixels]
 
     return pif_weight
+
+
+def _filter_PCA_pifs(reference_gimage, candidate_gimage, lim):
+    ''' Performs PCA analysis, on the valid pixels and filters according
+    to the distance from the principle eigenvector.
+    '''
+
+    logging.info('Pseudo invariant feature generation is using: Filtering '
+                 'using PCA.')
+
+    gimage.check_comparable([reference_gimage, candidate_gimage])
+
+    array_shape = candidate_gimage.bands[0].shape
+
+    # Only analyse valid pixels
+    alpha = numpy.ones(array_shape)
+    alpha[numpy.nonzero(candidate_gimage.alpha == 0)] = 0
+    alpha[numpy.nonzero(reference_gimage.alpha == 0)] = 0
+    alpha_vec = alpha.flatten()
+    valid_pixels_list = numpy.nonzero(alpha_vec != 0)
+
+    PIF_all = numpy.ones(array_shape)
+
+    # Per band analysis
+    for band_no in xrange(len(candidate_gimage.bands)):
```

Comment:
The PCA is done on a per band basis, so there isn't a strong reason to have all bands in memory at once. Your Vagrant VM is limited on memory, always less than available on you native system, so this will have a direct impact on VM performance.